client/src/message-handler.js

```diff
@@ -2,280 +2,523 @@ var SyncMessage = require('../../lib/syncmessage');
+    message.content = {path: path, type: type, checksums: checksums};
+    syncManager.send(message.stringify());
+  });
 }
 
 function handleRequest(syncManager, data) {
   var fs = syncManager.fs;
+  var rawFs = syncManager.rawFs;
   var sync = syncManager.sync;
-  var session = syncManager.session;
 
   function handleChecksumRequest() {
-    var srcList = session.srcList = data.content.srcList;
-    session.path = data.content.path;
-    fs.modifiedPath = null;
-    sync.onSyncing();
+    if(data.invalidContent(['type', 'sourceList'])) {
+      log.error('Path, type or source list not sent by server in handleChecksumRequest.', data);
+      return onError(syncManager, new Error('Server sent insufficient content'));
+    }
+
+    sendChecksums(syncManager, data.content.path, data.content.type, data.content.sourceList);
+  }
+
+  function handleDiffRequest() {
+    if(data.invalidContent(['type', 'checksums'])) {
+      log.warn(data, 'Upstream sync message received from the server without sufficient information in handleDiffRequest');
+      fs.delaySync(function(err, path) {
+        if(err) {
+          log.error(err, 'An error occured while updating paths to sync in handleDiffRequest');
+          return onError(syncManager, err);
+        }
+
+        log.info('Sync delayed for ' + path + ' in handleDiffRequest');
+        syncManager.currentSync = false;
+        syncManager.syncUpstream();
+      });
+      return;
+    }
+
+    var path = data.content.path;
+    var type = data.content.type;
+    var checksums = data.content.checksums;
+    var message;
+
+    rsync.diff(rawFs, path, checksums, rsyncOptions, function(err, diffs) {
+      if(err){
+        log.error(err, 'Error generating diffs in handleDiffRequest for ' + path);
+
+        fs.delaySync(function(delayErr, delayedPath) {
+          if(delayErr) {
+            log.error(err, 'Error updating paths to sync in handleDiffRequest after failing to generate diffs for ' + path);
+            return onError(syncManager, delayErr);
+          }
+
+          log.info('Sync delayed for ' + delayedPath + ' in handleDiffRequest');
+          syncManager.currentSync = false;
+          syncManager.syncUpstream();
+        });
+      } else {
+        message = SyncMessage.response.diffs;
+        message.content = {path: path, type: type, diffs: serializeDiff(diffs)};
+        syncManager.send(message.stringify());
+      }
+    });
+  }
+
+  function handleRenameRequest() {
+    if(data.invalidContent(['type', 'oldPath'])) {
+      log.error('Path, type or old path not sent by server in handleRenameRequest.', data);
+      return onError(syncManager, new Error('Server sent insufficient content'));
+    }
+
+    var path = data.content.path;
+    var oldPath = data.content.oldPath;
+    var type = data.content.type;
+    var message;
+
+    // If the server requests to downstream a path that is not in the
+    // root, ignore the downstream.
+    if(path.indexOf(fs.root) !== 0) {
```

Comment:
Hmm I think thats a bigger issue. The SyncFileSystem should handle that...which means we will have to add a wrapper to the Shell (I assume that those paths can only be used using filer's shell) that will Path.normalize() the given path.
I can file a separate bug on that?